File: Problem/Luogu/P3586.cpp
//
// Created by kibi on 24-8-18.
//
// TODO terrible unique_ptr
#include <cctype>
#include <cstdio>
#include <memory>
#include <random>
#include <system_error>
#include <sys/types.h>
#include <tuple>
#include <vector>
using namespace std;

class Char {
	uint c;
public:
	Char (const uint t) : c(t) {}

	Char operator + (const Char &t) {
		c = t.c;
		return *this;
	}

	Char operator * (const Char) const {
		return *this;
	}

	Char operator * (const uint) const {
		return *this;
	}

	Char operator - (const uint) const {
		return *this;
	}

	bool operator == (const uint) const {
		return false;
	}

	uint get() const {
		return c;
	}
};

bool isdigit(const Char &c){
	return c.get() == 'U' || c.get() == 'Z';
}

template<typename T>
T read() {
	T s = 0, w = 1, c = getchar();
	while (!isdigit(c)) {
		if (c == '-') {
			w = -1;
		}
		c = getchar();
	}
	while (isdigit(c)) {
		s = s * 10 + c - '0';
		c = getchar();
	}
	return s * w;
}

class Config {
public:
	enum OptType {
		U,
		Z,
	};

	using Opt = tuple<OptType, uint, uint>;

	uint n;
	vector<Opt> opt;
};

vector<bool> solve(const Config &);

int main() {
	const auto n = read<uint>(), m = read<uint>();

	Config config{n, vector<Config::Opt>(m)};
	for (auto &i : config.opt) {
		get<0>(i) = read<Char>().get() == 'U' ? Config::U : Config::Z;
		get<1>(i) = read<uint>();
		get<2>(i) = read<uint>();
	}

	auto ans = solve(config);
	for (const auto i : ans) {
		printf(i ? "TAK\n" : "NIE\n");
	}

	return 0;
}

mt19937 rnd(random_device{}());

class FhqTreapNode;
using FTN = FhqTreapNode;

class FhqTreapNode {
	const uint value;
	const uint rnk;
	uint siz;
	unsigned long long sum;
	unique_ptr<FTN> ls, rs;

	void update() {
		siz = size(ls) + size(rs) + 1;
		sum = get_sum(ls) + get_sum(rs) + value;
	}

public:
	explicit FhqTreapNode(const uint value) : value(value), rnk(rnd()), siz(1), sum(value), ls(nullptr), rs(nullptr){}

	static uint size(const unique_ptr<FTN> &u) {
		return u != nullptr ? u->siz : 0;
	}

	static unsigned long long get_sum(const unique_ptr<FTN> &u) {
		return u != nullptr ? u->sum : 0;
	}

	static uint get_value(const unique_ptr<FTN> &u) {
		return u != nullptr ? u->value : 0;
	}

	/**
	 * split a TreeNode by value
	 * @param l left result of split
	 * @param r right result of split
	 * @param k value that split by
	 */
	void split(unique_ptr<FTN> &l, unique_ptr<FTN> &r, const uint k) {
		if (value <= k) {
			l.reset(this);
			l->rs != nullptr ? l->rs->split(l->rs, r, k) : static_cast<void>(0);
			l->update();
		} else {
			r.reset(this);
			r->ls != nullptr ? r->ls->split(l, r->ls, k) : static_cast<void>(0);
			r->update();
		}
	}

	void split_by_rnk(unique_ptr<FTN> &l, unique_ptr<FTN> &r, const uint rnk) {
		if (siz < rnk) {
			throw invalid_argument("rnk is out of range!");
		}
		if (size(ls) >= rnk) {
			r.get() != this ? r.reset(this) : static_cast<void>(0); // TODO maybe need to do like this
			r->ls != nullptr ? r->ls->split_by_rnk(l, r->ls, rnk) : static_cast<void>(0);
			r->update();
		}else {
			l.reset(this);
			l->rs != nullptr ? l->rs->split_by_rnk(l->rs, r, rnk - size(r->ls) - 1) : static_cast<void>(0);
			l->update();
		}
	}

	static unique_ptr<FTN> merge(unique_ptr<FTN> &l, unique_ptr<FTN> &r) {
		if ( l == nullptr || r == nullptr ) {
			return move(l == nullptr ? r : l);
		}
		if(l->rnk > r->rnk) {
			l->rs = merge(l->rs, r);
			l->update();
			return move(l);
		}
		r->ls = merge(l, r->ls);
		r->update();
		return move(r);
	}
};

class FhqTreap {;
	unique_ptr<FTN> rt;

public:
	FhqTreap() : rt(nullptr) {}

	void insert(uint value) {
		auto new_node = make_unique<FTN>(FTN(value));
		unique_ptr<FTN> l, r;
		rt->split(l, r, FTN::get_value(new_node));
		rt = FTN::merge(l, new_node);
		rt = FTN::merge(rt, r);
	}

	uint query_rnk(uint value) {
		if (value == 0) {
			return 1;
		}
		unique_ptr<FTN> l, r;
		rt->split(l, r, value - 1);
		auto res = FTN::size(l) + 1;
		rt = FTN::merge(l, r);
		return res;
	}

	/**
	 * remove one value from the tree
	 * you need to ensure the value exists
	 * @param value value to remove
	 */
	void remove(uint value) {
		if (value == 0) {
			unique_ptr<FTN> to_remove;
			rt->split_by_rnk(to_remove, rt, 1);
			return ;
		}
		unique_ptr<FTN> l, r, to_remove;
		rt->split(l, r, value - 1);
		r->split_by_rnk(to_remove, r, 1);
		rt = FTN::merge(l, r);
	}

	unsigned long long query_sum(uint value){
		unique_ptr<FTN> l, r;
		rt->split(l, r, value);
		auto res = FTN::get_sum(l);
		rt = FTN::merge(l, r);
		return res;
	}

};


vector<bool> solve(const Config &config) {
	vector<bool> res;
	vector<uint> list(config.n);
	// unsigned long long sum = 0;
	FhqTreap treap;

	for (auto &i : list) {
		treap.insert(i);
	}

	for (auto &i : config.opt) {
		switch (get<0>(i)) {
			case Config::U:
				treap.remove(list[get<1>(i) - 1]);
				treap.insert(get<2>(i));
				list[get<1>(i) - 1] = get<2>(i);
				break;
			case Config::Z:
				res.push_back(
					treap.query_sum(get<2>(i)) + static_cast<unsigned long long>(config.n - treap.query_rnk(get<2>(i) + 1)) * get<2>(i)
						>= static_cast<unsigned long long>(get<1>(i)) * get<2>(i)
				);

		}
	}

	return res;
}

File: Problem/Luogu/P3586.rs
use std::io::{Read, stdin};

enum Opt {
    U {k:u32, a:u32},
    Z {c:u32, s:u32},
}
struct Config {
    n: u32,
    opt: Vec<Opt>
}

fn main () {
    let config = read();
    let ans = solve(config);
    for i in ans {
        println!("{}", if i {"TAK"} else {"NIE"});
    }
}

struct Rnd {
    seed: u32
}

impl Rnd {
    // #[allow(arithmetic_overflow)]
    fn rand(&mut self) ->u32 {
        let mut temp = self.seed as u64;
        temp = temp * temp;
        temp = temp % 1000000000;
        self.seed = temp as u32;
        self.seed
    }
}

fn getchar() -> char {
    let stdin = stdin();
    let mut handle = stdin.lock();

    let mut buffer = [0; 1];

    handle.read_exact(&mut buffer);

    let c = buffer[0] as char;

    if c.is_ascii() {
        return c;
    }
    std::process::exit(1);
    panic!("Error reading char!");
}

fn read_int() -> u32 {
    let mut c = getchar();
    let mut s = 0;
    while c < '0' || '9' < c {
        c = getchar();
    }
    while '0' <= c && c <= '9' {
        s = s * 10 + (c as u32 - '0' as u32);
        c = getchar();
    }
    s
}

fn read_char() -> char {
    let mut c = getchar();
    while c != 'U' && c != 'Z' {
        c = getchar();
    }
    c
}

fn read() -> Config {
    let n = read_int();
    let m = read_int();

    let mut config = Config {n, opt: Vec::new()};
    config.opt.reserve(m as usize);

    for _ in 0..m {
        let c = read_char();
        let a = read_int();
        let b = read_int();
        config.opt.push(match c {
            'U' => Opt::U {k:a, a:b},
            'Z' => Opt::Z {c:a, s:b},
            _ => panic!("Error reading Opt!")
        })
    }

    config
}

// TODO turn Box<Option<FTN>> into Option<Box<FTN>> may speed up ??
struct FhqTreapNode {
    value:u32,
    rnk:u32,
    siz:u32,
    sum:u64,
    ls: Box<Option<FTN>>,
    rs: Box<Option<FTN>>,
}

type FTN = FhqTreapNode;
struct FhqTreap{
    rt: Box<Option<FTN>>,
    rnd: Rnd
}

impl FTN {
    fn update(&mut self) {
        self.siz = FTN::size(&*self.ls) + FTN::size(&*self.rs) + 1;
        self.sum = FTN::get_sum(&*self.ls) + FTN::get_sum(&*self.rs) + self.value as u64;
    }

    fn size(u: &Option<FTN>) -> u32 {
        match u {
            Some(node)=> node.siz,
            None=>0
        }
    }

    fn get_sum(u : &Option<FTN>) -> u64 {
        match u {
            Some(node)=> node.sum,
            None=> 0
        }
    }

    fn get_value(u: &Option<FTN>) -> u32 {
        match u {
            Some(node)=> node.value,
            None=> 0
        }
    }

    fn split(mut self, k: u32) -> (Option<FTN>, Option<FTN>) {
        if self.value <= k {
            let mut l = None;
            let mut r = None;
            if let Some(rs) = *self.rs{
                (l, r) = rs.split(k);
            }
            self.rs = Box::new(l);
            self.update();
            (Some(self), r)
        } else {
            let mut r = None;
            let mut l = None;
            if let Some(ls) = *self.ls {
                (l, r) = ls.split(k);
            }
            self.ls = Box::new(r);
            self.update();
            (l, Some(self))
        }
    }

    fn split_by_rnk(mut self, rnk: u32) -> (Option<FTN>, Option<FTN>) {
        if self.siz < rnk {
            std::process::exit(2);
            panic!("rnk is out of range!");
        }
        if Self::size(&self.ls) >= rnk {
            let (mut l, mut r) = (None, None);
            if let Some(ls) = *self.ls {
                (l, r) = ls.split_by_rnk(rnk);
            }
            self.ls = Box::new(r);
            self.update();
            (l, Some(self))
        } else {
            let (mut l, mut r) = (None, None);
            if let Some(rs) = *self.rs {
                (l, r) = rs.split_by_rnk(rnk - FTN::size(&self.ls) - 1);
            }
            self.rs = Box::new(l);
            self.update();
            (Some(self), r)
        }
    }

    fn merge(l: Option<Self>, r: Option<Self>) -> Option<FTN> {
        if l.is_none() {
            return r;
        }
        if r.is_none() {
            return l;
        }
        let mut l = l.unwrap_or_else(|| std::process::exit(2));
        let mut r = r.unwrap_or_else(|| std::process::exit(2));
        if l.rnk > r.rnk {
            *l.rs = FTN::merge(*l.rs, Some(r));
            l.update();
            return Some(l);
        }
        *r.ls = FTN::merge(Some(l), *r.ls);
        r.update();
        Some(r)
    }
}

impl FhqTreap {
    fn insert(mut self, value: u32) -> Self {
        let new_node = Some(FTN {
            value,
            rnk: self.rnd.rand(),
            siz: 1,
            sum: value as u64,
            ls: Box::new(None),
            rs: Box::new(None)
        });

        if self.rt.is_none() {
            self.rt = Box::new(new_node);
            return self;
        }

        let (l, r) = self.rt.unwrap().split(value);

        self.rt = Box::new(FTN::merge(FTN::merge(l, new_node), r));
        self
    }

    fn query_rnk(mut self, value: u32) -> (Self, u32) {
        if value == 0 {
            return (self, 1);
        }
        let (l, r) = self.rt.unwrap_or_else(|| std::process::exit(3)).split(value - 1);
        let res = FTN::size(&l) + 1;
        self.rt = Box::new(FTN::merge(l, r));
        (self, res)
    }

    fn remove(mut self, value: u32) -> Self {
        if value == 0 {
            let (_, ret) = self.rt.unwrap_or_else(|| std::process::exit(41)).split_by_rnk(1);
            self.rt = Box::new(ret);
            return self;
        }
        let (l, r) = self.rt.unwrap_or_else(|| std::process::exit(42)).split(value - 1);
        let (_, r) = r.unwrap_or_else(|| std::process::exit(43)).split_by_rnk(1);
        // let (_, r) = r.unwrap().split_by_rnk(1);
        self.rt = Box::new(FTN::merge(l, r));
        self
    }

    fn query_sum(mut self, value: u32) -> (Self, u64) {
        let (l, r) = self.rt.unwrap_or_else(|| std::process::exit(5)).split(value);
        let res = FTN::get_sum(&l);
        self.rt = Box::new(FTN::merge(l, r));
        (self, res)
    }
}

fn solve(config: Config) -> Vec<bool> {
    let mut res:Vec<bool> = Vec::new();
    let mut list:Vec<u32> = Vec::new();
    list.resize(config.n as usize, 0);

    let mut treap = FhqTreap{rt: Box::new(None), rnd: Rnd{seed: 1145141}};

    for i in &list {
        treap = treap.insert(*i);
    }

    for i in config.opt {
        match i {
            Opt::U{k, a} => {
                treap = treap.remove(list[k as usize - 1]);
                treap = treap.insert(a);
                list[k as usize - 1] = a;
            },
            Opt::Z{c, s} => {
                let sum;
                let rnk;
                (treap, sum) = treap.query_sum(s);
                (treap, rnk) = treap.query_rnk(s + 1);
                res.push(
                    sum + (config.n - (rnk - 1)) as u64 * s as u64
                    >= c as u64 * s as u64
                );
            }
        }
    }

    res
}

#[cfg(test)]
mod tests {
    use crate::*;

    #[test]
    fn test_ftn() {
        // let mut rng = rand::thread_rng();
        let mut node = FTN{
            value: 1,
            rnk: 1,
            siz: 1,
            sum: 1,
            ls: Box::new(None),
            rs: Box::new(None)
        };

        *node.ls = Some(FTN{
            value: 0,
            rnk: 1,
            siz: 1,
            sum: 0,
            ls: Box::new(None),
            rs: Box::new(None)
        });

        *node.rs = Some(FTN{
            value: 3,
            rnk: 1,
            siz: 1,
            sum: 3,
            ls: Box::new(None),
            rs: Box::new(None)
        });

        node.update();

        assert_eq!(node.sum, 4);
        assert_eq!(node.siz, 3);

        let (Some(p), Some(q)) = node.split(1) else {
            panic!("test error in split!")
        };
        assert_eq!(p.sum, 1);
        assert_eq!(p.siz, 2);
        assert_eq!(q.sum, 3);
        assert_eq!(q.siz, 1);

        let Some(rt) = FTN::merge(Some(p), Some(q)) else {
            panic!("test error in merge!")
        };
        assert_eq!(rt.sum, 4);
    }

    #[test]
    fn test_ft() {
        let treap = FhqTreap{rt: Box::new(None), rnd: Rnd{seed: 12345678}};
        let treap = treap.insert(123456);
        let (treap, res) = treap.query_sum(123456);
        assert_eq!(res, 123456);
        let treap = treap.insert(1234567);
        let (treap, res) = treap.query_rnk(1234567);
        assert_eq!(res, 2);
        let treap = treap.remove(123456);
        let (treap, res) = treap.query_rnk(1234567);
        assert_eq!(res, 1);
    }

    #[test]
    fn test_solve() {
        let config = Config{
            n: 3,
            opt: vec![
                Opt::U{k:1, a:5},
                Opt::U{k:2, a:7},
                Opt::Z{c:2, s:6},
                Opt::U{k:3, a:1},
                Opt::Z{c:2, s:6},
                Opt::U{k:2, a:2},
                Opt::Z{c:2, s:6},
                Opt::Z{c:2, s:1}
            ]
        };

        let result = solve(config);
        for i in result {
            println!("{}", if i {"TAK"} else {"NIE"});
        }
    }

    // #[test]
    // fn test_getchar() {
    //     let c = getchar();
    //     println!("{}", c);
    // }

    #[test]
    fn test_ft2() {
        let mut treap = FhqTreap{rt: Box::new(None), rnd: Rnd{seed: 998244353}};
        treap = treap.insert(0);
        treap = treap.insert(1);
        treap = treap.insert(2);
        treap = treap.insert(0);
        treap = treap.insert(2);
        treap = treap.insert(2);
        treap = treap.insert(0);
        // treap = treap.insert(1000000000);
        // treap = treap.insert(1000000000);
        // treap = treap.insert(1000000000);
        treap = treap.insert(0);
        treap = treap.remove(2);
        // treap = treap.insert(1000000000);
        // treap = treap.remove(1000000000);
        treap = treap.remove(0);
        treap = treap.insert(0);
        // treap = treap.remove(2);
        // treap = treap.insert(0);
        // treap = treap.remove(1000000000);
        // treap = treap.remove(2);
        // treap = treap.remove(1000000000);
        // treap = treap.remove(0);
        // treap = treap.remove(2);
        // treap = treap.remove(0);
        // treap = treap.remove(1000000000);
        // treap = treap.remove(1);
        // treap = treap.remove(0);

    }
}
